keps/sig-node/0036-20190118-ForceEviction.md

```diff
@@ -0,0 +1,174 @@
+However, users often reported issues during the failure handling including old pods
+cannot be deleted correctly, new pods cannot be started correctly due to volume failing
+to detach from old node and attach to new node issues, or it takes a long time for new
+pods to start. The main reason for this problem is due to the fact that Pod deletion requires
+kubelet's acknowlegement. When a Pod is eviceted (deleted) from a node, the deletion timestamp
+is first set on the Pod. However the Pod will not be actually deleted from API server until kubelet 
+finishes cleaning up all the resources of this Pod and kills the running containers. If the node
+is in a failed state and kubelet cannot respond, Pod will be stuck in a teminating state forever.
+In addtion, the containers status for the Pod are still running since kubelet cannot update it,
+which blocks attach_detach controller to detach volumes. The current taint-based eviction actually
+cannot give any benefit in the sense that Pod can only be succesfully evicted after Kubelet comes
+back healthy, in which case there is no need of Pod eviction at all.
+
+
+### Goals
+
+This proposal is trying to investigate the failure handling issues for pods
+and their volumes with a goal of improving system reliability and failure recovery capabilities.
+Among different kinds of failures, node shut down is the main user case this proposal tries
+to address. In this case, we are sure that all containers are stopped, and all mounts are gone
+so that it is safe to take some recovery actions more aggresively without waiting for kubelet correspondence.  
+
+to delete the Pods from API server without waiting for kubelet's acknoledgement.
+Volumes can also be safely detached since mounts no longer exist.
+
+### Non-Goals
+
+This proposal is not trying to address the recovery issue for all types of failures. For example for network
+partition issue, because this is no source of truth about whether Pods are running or not, it is not possible
+for controller decide what recovery actions should take based on the node status.
+
+## Proposal
+
+Currently, Kubernetes already supports four types of taint effects: 
+- NoSchedule: Pods that do not tolerate this taint are not scheduled on the node.
+- PreferNoSchedule: Kubernetes avoids scheduling Pods that do not tolerate this taint onto the node.
+- NoExecute: Pod is evicted from the node if it is already running on the node, and is not scheduled
+onto the node if it is not yet running on the node.
+
+We propose to add a new effect, TaintEffectForceEviction. For a Pod that cannot tolerant this taint,
+it will be force deleted from the node with zero grace period which means that Pod can be deleted immediately
+without waiting for Kubelet to responde. One important use case of this taint is node shut down case. If node
+controller or some external controller detact that node is powered off, the controller could add the following
+taint to the node.
+
+
+```
+taint := &v1.Taint{
```

Comment:
We should state clearly that daemonSets should tolerate this new taint cc @kubernetes/sig-apps-misc

Reply:
This is suspect. What should be done with existing daemonsets and daemonset pods? Any time we have to make existing workload controllers change in response to a new node taint or lifecycle change, we are essentially recognizing a breaking change in how pod-creating controllers interact with the node. What are the implications for pod-creating controllers that are out of tree that we can't sweep/update?

Reply:
@liggitt - The implication to me are two:

- daemonset controller mostly, should ensure that daemonsets pods tolerate this new taint when reconciling.

- out-of-tree workload controllers: I agree with you, we implicitly break pod-creating controllers that did not use a wildcard toleration. One way that we discussed together with @jingxu97 is to feature-gate this in order to give enough visibility for these controllers.

Reply:
@liggitt - also please note kubernetes/kubernetes#60009 introduced a noSchedule taint called `node.cloudprovider.kubernetes.io/shutdow`. We should reflect on the KEP how to deal with it cc @jingxu97

Reply:
> * daemonset controller mostly, should ensure that daemonsets pods tolerate this new taint when reconciling.

One of the main reasons to have the daemonset run its pods through the scheduler was to honor scheduling constructs like resource availability and taints/tolerations. Having the DaemonSet controller try to adjust pods to auto-tolerate taints steals power from the API (and also is a sign that the new taint is likely a compatibility problem). Can you describe why daemonset pods are special and should not be subject to deletion in the circumstances that would cause a node to be tainted this way?

> * out-of-tree workload controllers: I agree with you, we implicitly break pod-creating controllers that did not use a wildcard toleration.

I have not dug into this proposal enough to understand all the problems it is addressing, but on the surface, that sounds very problematic.

Reply:
talked to @yujuhong about deamonSet. Some deamonSet might run important services for the cluster should tolerate all taints. For example, GKE put a toleration with wild card for some deamonSet it is running so that it can tolerate all taints. I think it depends on the system admin on how to set up the toleration for some deamonSets. For this feature, we don't need to adjust deamonSet pods to auto-tolerate the taint, I think

Reply:
My experience with OpenShift is vast majority of daemonset pods tolerate all taints.  Of course, GKE and OpenShift alone is representing how a service provider is using the feature but not necessarily the end users.